skunkworks_crow/src/main/java/org/odk/share/views/ui/bluetooth/BluetoothListAdapter.java

```diff
@@ -0,0 +1,134 @@
+package org.odk.share.views.ui.bluetooth;
+
+import android.bluetooth.BluetoothAdapter;
+import android.bluetooth.BluetoothDevice;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.TextView;
+
+import androidx.annotation.NonNull;
+import androidx.recyclerview.widget.RecyclerView;
+
+import org.odk.share.R;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import butterknife.BindView;
+import butterknife.ButterKnife;
+
+
+/**
+ * BluetoothListAdapter: {@link RecyclerView.Adapter} for {@link BluetoothDevice}.
+ *
+ * @author huangyz0918 (huangyz0918@gmail.com)
+ * @since 12/06/2019
+ */
+public class BluetoothListAdapter extends RecyclerView.Adapter<BluetoothListAdapter.ViewHolder> {
+
+    private List<BluetoothDevice> bluetoothDeviceList;
+    private OnDeviceClickListener btDeviceClickListener;
+
+    public BluetoothListAdapter(OnDeviceClickListener btDeviceClickListener) {
+        this.btDeviceClickListener = btDeviceClickListener;
+        this.bluetoothDeviceList = new ArrayList<>();
+    }
+
+    /**
+     * Methods for updating data.
+     */
+    public void setDevices(List<BluetoothDevice> dateSet) {
+        this.bluetoothDeviceList = dateSet;
+    }
+
+    public void addDevice(BluetoothDevice bluetoothDevice) {
+        if (!bluetoothDeviceList.contains(bluetoothDevice)) {
+            this.bluetoothDeviceList.add(bluetoothDevice);
+        }
+        notifyDataSetChanged();
+    }
+
+    @Override
+    public void onBindViewHolder(@NonNull final ViewHolder viewHolder, int position) {
+        BluetoothDevice device = bluetoothDeviceList.get(position);
+        String deviceName = device.getName();
+        String deviceAddress = device.getAddress();
+        int deviceBondState = device.getBondState();
+        viewHolder.deviceName.setText(deviceName == null ? viewHolder.itemView.getResources().getString(R.string.bluetooth_instance_name_default) : deviceName);
+        viewHolder.deviceAddress.setText(String.format("%s (%s)", deviceAddress, deviceBondState == 10 ? "unpaired" : "paired"));
+    }
+
+    @Override
+    public int getItemCount() {
+        return bluetoothDeviceList.size();
+    }
+
+    @NonNull
+    @Override
+    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_bt_device, null);
+        return new ViewHolder(view);
+    }
+
+
+    /**
+     * rescan the bluetooth devices and update the list.
+     */
+    public void rescan() {
```

Comment:
rescan logic should not be present in adpater class same for below method `addBoundedDevices`

Reply:
See updated.